music_assistant/server/providers/snapcast/__init__.py

```diff
@@ -0,0 +1,305 @@
+        def on_progress(progress: Progress):
+            player.state = PlayerState.PLAYING
+            self.mass.players.register_or_update(player)
+
+        @ffmpeg.on("completed")
+        async def on_completed():
+            player.state = PlayerState.IDLE
+            self.mass.players.register_or_update(player)
+
+        @ffmpeg.on("terminated")
+        async def on_terminated():
+            player.state = PlayerState.IDLE
+            self.mass.players.register_or_update(player)
+
+    async def cmd_stop(self, player_id: str) -> None:
+        """Send STOP command to given player."""
+        player = self.mass.players.get(player_id, raise_unavailable=False)
+        if player.state != PlayerState.IDLE:
+            stream = self._get_snapstream(player_id)
+            if hasattr(stream, "ffmpeg"):
+                try:
+                    stream.ffmpeg.terminate()
+                    self.logger.debug("ffmpeg player stopped")
+                except FFmpegError:
+                    self.logger.debug("Fail to stop ffmpeg player")
+
+    async def cmd_pause(self, player_id: str) -> None:
+        """Send PAUSE command to given player."""
+        await self.cmd_stop(player_id)
+
+    async def cmd_volume_mute(self, player_id, muted):
+        """Send MUTE command to given player."""
+        self.mass.create_task(self._snapserver.client(player_id).set_muted(muted))
+
+    async def _remove_stream(self, stream_id):
+        self.mass.create_task(self._server.stream_remove_stream(stream_id))
+
+    def _snapclient_get_group_clients_identifiers(self, player_id):
+        group = self._get_snapgroup(player_id)
+        return [ele for ele in group.clients if ele != player_id]
+
+    async def cmd_sync(self, player_id: str, target_player: str) -> None:
+        """Sync Snapcast player."""
+        child_player = self.mass.players.get(player_id)
+        assert child_player  # guard
+        parent_player = self.mass.players.get(target_player)
+        assert parent_player  # guard
+        # always make sure that the parent player is part of the sync group
+        # parent_player.group_childs.add(parent_player.player_id)
+        parent_player.group_childs.add(child_player.player_id)
+        child_player.synced_to = parent_player.player_id
+
+        group = self._get_snapgroup(target_player)
+        self.mass.create_task(group.add_client(player_id))
+
+        self.mass.players.update(child_player.player_id)
+        self.mass.players.update(parent_player.player_id)
+
+    async def cmd_unsync(self, player_id: str) -> None:
+        """Unsync Snapcast player."""
+        group = self._get_snapgroup(player_id)
+        await group.remove_client(player_id)
+        group = self._get_snapgroup(player_id)
+        stream_id = await self._get_empty_stream(player_id)
+        await group.set_stream(stream_id)
+        self._handle_update()
+
+    def _get_snapgroup(self, player_id):
+        client = self._snapserver.client(player_id)
```

Comment:
All methods below are mnissing docstrings

Reply:
and missing return annotations

Reply:
Ok